src/appleseed/foundation/array/arrayallocator.h

```diff
@@ -0,0 +1,168 @@
+// THE SOFTWARE.
+//
+
+#pragma once
+
+// appleseed.foundation headers.
+#include "foundation/utility/memory.h"
+
+// Standard headers.
+#include <cstddef>
+#include <limits>
+#include <new>
+
+namespace foundation
+{
+
+//
+// A standard-conformant allocator allocating aligned, padded memory.
+//
+
+template <typename T>
+class ArrayAllocator
+{
+  public:
+    typedef T                   value_type;
+    typedef value_type*         pointer;
+    typedef const value_type*   const_pointer;
+    typedef value_type&         reference;
+    typedef const value_type&   const_reference;
+    typedef size_t              size_type;
+    typedef std::ptrdiff_t      difference_type;
+
+    // Alignment and padding must be powers of 2.
+    static const size_t Alignment = 16;
+    static const size_t Padding = 16;
+
+    template <typename U>
+    struct rebind
+    {
+        typedef ArrayAllocator<U> other;
+    };
+
+    ArrayAllocator() = default;
+
+    template <typename U>
+    ArrayAllocator(const ArrayAllocator<U>&)
+    {
+    }
+
+    bool operator==(const ArrayAllocator<T>&) const
+    {
+        return true;
+    }
+
+    template <typename U>
+    bool operator==(const ArrayAllocator<U>&) const
+    {
+        return false;
+    }
+
+    template <typename U>
+    bool operator!=(const ArrayAllocator<U>&) const
+    {
+        return true;
+    }
+
+    pointer address(reference x) const
+    {
+        return &x;
+    }
+
+    const_pointer address(const_reference x) const
+    {
+        return &x;
+    }
+
+    pointer allocate(size_type n, const_pointer hint = nullptr)
+    {
+        if (n == 0)
+            return nullptr;
+
+        const size_type padded_size = round_up_size(n * sizeof(T));
+        pointer p = static_cast<pointer>(aligned_malloc(padded_size, Alignment));
+
+        if (p == nullptr)
+             throw std::bad_alloc();
+
+        return p;
+    }
+
+    void deallocate(pointer p, size_type n)
+    {
+        if (p)
+            aligned_free(p);
+    }
+
+    size_type max_size() const
+    {
+        return std::numeric_limits<size_type>::max() / sizeof(T);
+    }
+
+    void construct(pointer p, const_reference x)
+    {
+        new(p) value_type(x);
+    }
+
+    void destroy(pointer p)
+    {
+        p->~value_type();
+    }
+
+  private:
+    size_type round_up_size(const size_type size)
+    {
+        return (size + Padding - 1) & -Padding;
+    }
+};
+
+// A partial specialization for the void value type is required for rebinding
+// to another, different value type.
+template <>
+class ArrayAllocator<void>
+{
+  public:
+    typedef void                value_type;
+    typedef value_type*         pointer;
+    typedef const value_type*   const_pointer;
+
+    template <typename U>
+    struct rebind
+    {
+        typedef ArrayAllocator<U> other;
+    };
+
+    ArrayAllocator() = default;
+
+    template <typename U>
+    ArrayAllocator(const ArrayAllocator<U>&)
```

Comment:
Why no `= default` here?

Reply:
Does not work for templated copy constructors.

Reply:
Ok.